### ISSUE TITLE
64-bit precision problem in grad 

### ISSUE THREAD

**Issue by mgbukov** [google/jax#3423]
I encountered a precision issue using `grad` on a custom function with 64-bit precision. Here's a minimal script to reproduce (mac osx, jax=0.1.62, jaxlib=0.1.43):

```
from jax.config import config
config.update("jax_enable_x64", True)

from jax import grad
import jax.numpy as jnp


def cosh_real_faulty(x):
    Re_x, Im_x = x
    # Cosh[a + I b] = Cos[b] Cosh[a] + I Sin[b] Sinh[a]
    Re_z = jnp.cos(Im_x)*jnp.cosh(Re_x)
    Im_z = jnp.sin(Im_x)*jnp.sinh(Re_x)
    return Re_z+Im_z


def cosh_cpx(x):
    x_real, x_imag = x
    z=jnp.cosh(x_real+1j*x_imag)
    return z.real+z.imag


def exact_grad(x):
	x_real, x_imag = x
	z=jnp.sinh(x_real+1j*x_imag)
	return z.real+z.imag

d_cosh_cpx=grad(cosh_cpx, holomorphic=False)
d_cosh_real_faulty=grad(cosh_real_faulty, holomorphic=False)


z=(0.001, 0.002)
print('numpy      {0:0.15f}'.format(d_cosh_cpx(z)[0]) )
print('custom     {0:0.15f}'.format(d_cosh_real_faulty(z)[0]) )
print('exact cosh {0:0.15f}'.format(exact_grad(z)) )

print()
z=(0.001, 0.001)
print('numpy      {0:0.15f}'.format(d_cosh_cpx(z)[0]) )
print('custom     {0:0.15f}'.format(d_cosh_real_faulty(z)[0]) )
print('exact cosh {0:0.15f}'.format(exact_grad(z)) )


print()
z=(0.01, 0.01)
print('numpy      {0:0.15f}'.format(d_cosh_cpx(z)[0]) )
print('custom     {0:0.15f}'.format(d_cosh_real_faulty(z)[0]) )
print('exact cosh {0:0.15f}'.format(exact_grad(z)) )
```

Output:

```
(jax) [mgbukov@mgbukovs-MBP cosh_precision_bug]$ python cosh_precision_bug.py 
/Users/mgbukov/miniconda3/envs/jax/lib/python3.7/site-packages/jax/lib/xla_bridge.py:123: UserWarning: No GPU/TPU found, falling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
/Users/mgbukov/miniconda3/envs/jax/lib/python3.7/site-packages/jax/lax/lax.py:1975: ComplexWarning: Casting complex values to real discards the imaginary part
  lambda t, new_dtype, old_dtype: [convert_element_type(t, old_dtype)])
numpy      0.002999997833333
custom     0.002999997666792
exact cosh 0.002999997833333

numpy      0.002000000000000
custom     0.001999999833458
exact cosh 0.002000000000000

numpy      0.019999999993333
custom     0.019999999993333
exact cosh 0.019999999993333
```

Notice the `ComplexWarning` above; not sure if it is related though.

**Comment by shoyer** [comment]
Could you kindly explain the specific issue here?

Note that in general, you cannot expect exact equality from two different ways of doing the same computation with floating point numbers. Here it looks like the two different ways of computing the derivative of cosh agree up to about 1e-10.

**Comment by mgbukov** [comment]
Sure! 

What I have are two functions: 
(i) the manually written`cosh_real_faulty` which computes the `cosh(z)` of a complex number by decomposing it into its real and imaginary part, and 
(ii) the numpy-based version `cosh_cpx`. 

The goal is to compute the gradient of `Re(cosh(z)) + Im(cosh(z))` [this is a minimal example coming from a more complex code]. I compute the gradients in two ways: 
(1) using `grad` on the two functions `cosh_real_faulty` and `cosh(z)`, and 
(2) analytically, see `exact_grad`.

Last, I compare the results on three different complex numbers `z` (defined as a tuple of real and imaginary parts).

I print the results to the 15-digit, and I was expecting to see an agreement between the results of the three functions up to the 15th digit because I use 64-bit precision. This should be true for every one of the three tests separately. However, the agreement appears to be only the case in the last test. 

@shoyer yes, we do have an agreement up to 1E-10 but I believe the agreement should be up to 1E-15 (like it is in the last test). Notice that when I use the numpy-only functions -- `cosh_cpx` -- I do get an agreement up to 1E-15 as compared to the exact derivative. It looks to me that some precision might be lost under the hood here, that's why I decided to report this.

**Comment by shoyer** [comment]
I guess I'm not too surprised here. You are comparing derivatives computed from the much simpler function `cosh_cpx`, a primitive with has a gradient rule directly defined (`jnp.cosh`) to your compound function `cosh_real_faulty`.

If you're curious how these computations work, I would suggest printing the JAXprs:
```
z=(0.001, 0.002)
print(jax.make_jaxpr(d_cosh_cpx)(z))
print(jax.make_jaxpr(d_cosh_real_faulty)(z))
```
which outputs:
```
{ lambda  ; a b.
  let c = convert_element_type[ new_dtype=complex128
                                old_dtype=float64 ] a
      d = convert_element_type[ new_dtype=complex128
                                old_dtype=float64 ] b
      e = mul d 1j
      f = add c e
      g = sinh f
      h = mul (1-1j) g
      i = convert_element_type[ new_dtype=float64
                                old_dtype=complex128 ] h
      j = mul h 1j
      k = convert_element_type[ new_dtype=float64
                                old_dtype=complex128 ] j
  in (i, k) }
{ lambda  ; a b.
  let c = sin b
      d = mul 1.0 c
      e = cosh a
      f = mul d e
      g = cos b
      h = mul 1.0 g
      i = sinh a
      j = mul h i
      k = add_any f j
      l = sinh a
      m = mul 1.0 l
      n = cos b
      o = mul m n
      p = cosh a
      q = mul 1.0 p
      r = neg q
      s = sin b
      t = mul r s
      u = add_any o t
  in (k, u) }
```

As you can see, `d_cosh_real_faulty` involves far more trigonometric operations. So it's not so surprising to me that the error accumulates more.

I suspect the second calculation may also rely upon subtracting numbers that are very close to each other, which could cause some of the loss of precision.

**Comment by mgbukov** [comment]
@shoyer  I think you're right -- it must be a machine precision issue. I defined the analytical gradient function separately, using the trigonometric identities

```
def exact_grad2(x):
    Re_x, Im_x = x
    # Sinh[a + I b] = Sin[a] Cosh[b] + I Cos[a] Sinh[b]
    Re_z = jnp.sin(Re_x)*jnp.cosh(Im_x)
    Im_z = jnp.cos(Re_x)*jnp.sinh(Im_x)
    return Re_z+Im_z
```

and I got a difference in the same digit. This one does not really involve the `grad` function so it must be the order of operations performed under the hood.
